### ISSUE TITLE
04_02 PreProcessing being extremely slow in comparison to 04_01

### ISSUE THREAD

**Issue by variableVG** [FreshJonas/NLP-on-Movie-Plots#3]
(Jonas) 
I am currently trying to implement a custom pre processing to the PyTorch Classifcation seen in 04_01. The current version in 04_02 works partly. I am able to use the model (takes 10 times longer as in 04_01) but I am not able to verify the model by using the testing data.

So there are two issues.

    PreProcessing being extremely slow in comparison to 04_01
    PreProcessed Version does not work for validating Model with Testing Data

**Comment by variableVG** [comment]
First: What would be an acceptable time for you? How long does it take (approximately) in your computer to run each of the models? I used google gpus to run your code and I get this results: 

Model in 04_01
![image](https://user-images.githubusercontent.com/58001220/200567053-ef6c5054-e29e-4c6b-be54-0cf5dc430723.png)


Model in 04_02
![image](https://user-images.githubusercontent.com/58001220/200567112-0a62aa30-2541-4f90-90ba-415c8364061c.png)

Would that be (2minutes) an acceptable time for you?

**Comment by variableVG** [comment]
**_Second:_** I realize you are using this line of code: `device = torch.device("cuda" if torch.cuda.is_available() else "cpu")`

With  cuda you can use high performance GPUs, but I think for that you need to have the capacity to produce GPUs in your computer. If you want, you can try the code in Google Colabs and use the GPUs of Google Colabs. You need to "activate" the GPUs in the following way: 

In the upper menu: Edit -> Notebook Settings 
![image](https://user-images.githubusercontent.com/58001220/200568677-80c5c62c-87d6-4f35-bc9d-b7f1ceb8bbd3.png)


Than select GPUs 
![image](https://user-images.githubusercontent.com/58001220/200569052-26534702-6abe-4237-907b-ce48486f5037.png)


In my case it increases the performance by double. That means, that 04_02 would take "just" 5 times as much as 04_01

**Comment by variableVG** [comment]
Maybe it is a bit of stupid question, but do you really tokenize in 04_02? It seems like you do according to the test sample you use, but I could not locate in the code exactly where are you doing tokenization. An then in the train_value_tuples that you use in the model, if you print them, they come as normal text. Is that really correct? Would you not want to pass the tokens instead of the whole text? 

![image](https://user-images.githubusercontent.com/58001220/200572820-41896d75-e8d4-48e6-881a-e845da60cedb.png)

**Comment by variableVG** [comment]
I will need a bit more of time, to understand completely the code, and maybe this is a bit of stupid question, but what are the differences between 02 and 01? I see that 02 has a lower learning rate, but other than that?

**Comment by FreshJonas** [comment]
The thing is that the data pre processing  with pytorch is a lot more complex then that. You first define a text_pipeline as well as a label_pipeline. Here lie the differences between 04_01 and 04_02

**Comment by FreshJonas** [comment]
![image](https://user-images.githubusercontent.com/76001421/200576077-9c783689-a067-4764-9592-84df8bcd233e.png)
This is what happens in the 04_01 pipeline under the hood

**Comment by FreshJonas** [comment]
![image](https://user-images.githubusercontent.com/76001421/200576260-b268bbb0-c338-42d3-b817-ca008372461f.png)
This is what happens in the 04_02 pipeline under the hood

**Comment by FreshJonas** [comment]
The time difference i am concerned about is the one when actually running the model. Here 04_02 is TEN TIMES slower than 04_01. The question is what I did wrong in the implementation of the 04_02 text_pipeline
